From: =?UTF-8?q?Guido=20G=C3=BCnther?= <agx@sigxcpu.org>
Date: Mon, 26 Jul 2010 17:55:40 +0200
Subject: [PATCH] Use g_strjoinv instead of g_strjoin

---
 src/cups.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/cups.c b/src/cups.c
index e439051..4df9945 100644
--- a/src/cups.c
+++ b/src/cups.c
@@ -1227,12 +1227,12 @@ _cph_cups_devices_get_14 (CphCups            *cups,
                 timeout_param = timeout;
 
         if (include_schemes && len_include > 0)
-                include_schemes_param = g_strjoin (",", include_schemes);
+                include_schemes_param = g_strjoinv (",", include_schemes);
         else
                 include_schemes_param = g_strdup (CUPS_INCLUDE_ALL);
 
         if (exclude_schemes && len_exclude > 0)
-                exclude_schemes_param = g_strjoin (",", exclude_schemes);
+                exclude_schemes_param = g_strjoinv (",", exclude_schemes);
         else
                 exclude_schemes_param = g_strdup (CUPS_EXCLUDE_NONE);
 
-- 
